src/logic/__tests__/startAccountMigration.test.js

```diff
@@ -0,0 +1,107 @@
+import {Observable} from 'rxjs';
+
+import startAccountMigration from '../startAccountMigration';
+import {
+  accountMigrationComplete,
+  accountMigrationError,
+  accountMigrationUndoPeriodExpired,
+} from '../../actions/user';
+import {migrateAccount} from '../../clients/firebase';
+import {getCurrentAccountMigration} from '../../selectors';
+import {bugsnagClient} from '../../util/bugsnag';
+
+import {
+  credentialFactory,
+  firebaseErrorFactory,
+  userFactory,
+} from '@factories/clients/firebase';
+
+import {
+  githubRepositoryFactory,
+} from '@factories/clients/github';
+
+
+jest.mock('../../actions/user.js');
+jest.mock('../../clients/firebase.js');
+jest.mock('../../selectors');
+jest.mock('../../util/bugsnag.js');
+
+const getState = jest.fn();
+const dispatch = jest.fn();
+const done = jest.fn();
+
+describe('startAccountMigration', () => {
+  test('not dismissed during undo period, successful migration', async() => {
+    const mockCredential = credentialFactory.build();
+    getCurrentAccountMigration.mockResolvedValue({
+      firebaseCredential: mockCredential,
+    });
```

Comment:
Rather than mocking the behavior of the selector, let’s just give it a `state` object that contains the data in question. That way, if we decide to change how selectors are used in the future in the logic (in a way that still produces the correct outcome), the test won’t break.

The preferred way to create a state object is to just run a series of actions through the reducer and grab the output. See the [console reducer test](https://github.com/popcodeorg/popcode/blob/master/src/reducers/__tests__/console.test.jsx#L143-L149) for a good example. In this case though we will want to run the actions through the root reducer, rather than just the reducer for the console namespace.

Reply:
Okay looked into this, but got a stuck when running the reducer.

The `action` parameter is coming in null for me:
https://github.com/popcodeorg/popcode/blob/master/src/reducers/user.js#L64

My reducer code looks something like this:
```
import reduce from 'lodash-es/reduce';
import reduceRoot from '../../reducers/index';
...
// inside test
const state= applyActions(
    startAccountMigrationAction,
    accountMigrationUndoPeriodExpired,
    accountMigrationNeeded({credential: mockCredential}),
);
...
function applyActions(...actions) {
  return reduce(
    actions,
    (state, action) => reduceRoot(state, action),
    undefined,
  );
}
```

Reply:
Hmm well one thing that jumps out is you’ve gotta call all the action creators, e.g. `startAccountMigrationAction()` not `startAccountMigrationAction`